src/Catty/ViewController/Download/StoreProjectDownloader.swift

```diff
@@ -183,3 +185,41 @@ enum ProjectType {
     case mostViewed
     case mostRecent
 }
+
+extension StoreProjectDownloader {
+    func download(project: StoreProject, completion: @escaping (Data?, StoreProjectDownloaderError?) -> Void, progression: ((Float) -> Void)?) {
+        guard let indexURL = URL(string: "https://share.catrob.at/app/download/\(project.projectId).catrobat") else { return }
+
+        let task = self.session.dataTask(with: URLRequest(url: indexURL)) { data, response, error in
+            let handleDataTaskCompletion: (Data?, URLResponse?, Error?) -> (projectData: Data?, error: StoreProjectDownloaderError?)
+            handleDataTaskCompletion = { data, response, error in
+                guard let response = response as? HTTPURLResponse else { return (nil, .unexpectedError) }
+
+                if let error = error as NSError?, error.domain == NSURLErrorDomain && error.code == NSURLErrorTimedOut {
+                                   return (nil, .timeout)
+                               }
+                guard let data = data, response.statusCode == 200, error == nil else {
+                    return (nil, .request(error: error, statusCode: response.statusCode)) }
+
+                if let error = error {
+                     return (nil, .parse(error: error))
+                }
+
+                return (data, nil)
+            }
+
+            let result = handleDataTaskCompletion(data, response, error)
+            DispatchQueue.main.async {
+                completion(result.projectData, result.error)
+            }
+        }
+
+        if let progression = progression {
+            observer = task.observe(\.countOfBytesReceived) { progress, _ in
+                progression(Float(progress.countOfBytesReceived) / Float(progress.countOfBytesExpectedToReceive))
```

Comment:
Nice! I was just wondering whether it would make sense to dispatch this block on the main queue, like the completion block? What do you think?

Reply:
Since, I was able to get proper progress from the progression block without moving it into the main thread and the UI is aslo behaving as expected so I don't think it is needed. But if there is any specific requirement that I am missing please tell me I would make the changes.

Reply:
I just saw that in the code base, in the current progress indicator, the UI changes for this indicator are dispatched on the main queue. In case this is necessary, which I think it will be, we could also make this dispatch in the completion block, so not a big deal and both methods are alright for me -- just found it a bit confusing that the first completion block is dispatched but the second not.